src/BreakpointViewer.cpp

```diff
@@ -139,18 +158,22 @@ void BreakpointViewer::_createBreakpoint(BreakpointRef::Type type, int row)
 	const QString cmdStr = Breakpoints::createSetCommand(wtype, range, slot, segment, condition);
 
 	auto* command = new Command(cmdStr,
-		[this, type, row] (const QString& id) {
-			setTextField(type, row, ID, id);
+		[=] (const QString& id) {
+			auto tableIndex = getTableIndexByRow(type, row);
+			assert(tableIndex);
 
-			BreakpointRef ref{type, id, row, -1};
-			if (!connectBreakpointID(id, ref)) {
-				setBreakpointChecked(type, row, Qt::Unchecked);
-				_handleKeyAlreadyExists();
-				return;
+			// remove old BreakpointRef
+			if (auto* ref = findBreakpointRef(type, row)) {
+				maps[type].erase(ref->id);
 			}
 
-			auto sa = ScopedAssign(selfUpdating, true);
-			emit contentsUpdated(false);
+			// ref doesn't exist, so it must be a new breakpoint
+			BreakpointRef newRef = {type, id, *tableIndex, -1};
+			addBreakpointRef(id, newRef);
+
+			// update breakpoint id
+			setTextField(type, row, ID, id);
+			selfUpdating = false;
 		},
 		[this] (const QString& error) { _handleSyncError(error); }
 	);
```

Comment:
It seems I did a review of an older version of this PR already 9(!) months ago?
Unfortunately I don't remember the details of the code, nor the review :(
And because you rebased (and then force-pushed) your patch, it's not clear which parts of the code are new or which were already part of the old version of the patch :(

Anyway, I didn't do a full in-depth review of your PR. I trust you addressed all (serious) problems I reported last time.

I did do a shallow review of the code. It mostly looks fine.
But here is something I have a question about:

In the successful-command-callback you reset `selfUpdating` to false. Therefor I assume `_createBreakpoint()` must be called with `selfUpdating` set to `true`? If this assumption is correct, then it might be a good idea to 'document' this via an `assert(selfUpdating)` statement at the start of this method.

In case of a command-failure you call `_handleSyncError()`, but that code path does not reset `selfUpdating` to `false`. Is that a bug?

I would like to ask you to go over your patch and check usages of `selfUpdating`, possibly document assumptions about this variable via `assert()` statements. After that I'll merge your PR.  (Possibly all is already correct, but it's not obvious to me).

Reply:
`selfUpdate` is by design almost always `false`. It was mostly used to stop `DebuggerForm` from calling `BreakpointViewer::sink` when the update came from `BeakpointViewer` itself and thus avoid the impending infinite loop of the cascading events. But this specific problem with `selfUpdating` here is due to `replaceBreakpoint` erasing the table row, since it would remove the breakpoint along with some useful information from the table. If we skip the `sync` called after the replacement (removal) of the breakpoint, we can reuse the table row on the next call to `createBreakpoint`. Setting `selfUpdating` to `false` afterwards is just restoring it to its default value.

I checked the behaviour of this code in both debug enabled/disabled mode and it seems to work OK for both.

Reply:
>  And because you rebased (and then force-pushed) your patch, it's not clear which parts of the code are new or which were already part of the old version of the patch :(

This happened because I created this PR on master, which I must remember not to do again. Then I closed it and I synchronised my rep because of other PRs and then the code in this branch was forever lost.

Reply:
Have I missed one? I believe all paths are already addressed. And `removeBreakpoint` doesn't change `selfUpdating` to `true` anywhere.

Reply:
Maybe I'm wrong. But let me explain my reasoning and assumptions:
(All line numbers refer to the new version of the code)

* On line 157 you create a new Command. This needs two callbacks, one for success, and one for error.
* On line 174 on the success-path you reset `selfUpdating`. Therefor my assumption was that you should do the same on the error-path.
* I checked how _createBreakPoint() can get called. It only seems to be called from createBreakPoint() (on line 104). That function is called from line 199 and 492).
* At least in case of the former, on line 193, before calling createBreakPoint() you set `selfUpdating=true`, and a bit lower on the error-handling path on line 203, you reset it to `false`. This strengthens my believe that on the error-path on line 178 you should do the same. ..... Euhm, you already do? Ah, but you changed that only recently.

Ok, seems like you fixed my concern :-)
I'll pull your PR later today when I'm back at home.